## 1. Problem Statement

We have the dataset of 1000 popular movies on IMDB in the last 10 years ie, 2006-2016.

This Exploratory Data Analysis is to practice Python skills learned till now on a structured data set including loading, inspecting, wrangling, exploring, and drawing conclusions from data.

## 2. Data Loading and Description

This dataset includes 1000 observations of 12 columns. The data fields included are:

| Column Name         | Description                                                                              |
| -------------       |:-------------                                                                           :| 
| Rank                | Movie rank order                                                                         | 
| Title               | The title of the film                                                                    |  
| Genre               | A comma-separated list of genres used to classify the film                               | 
| Description         | Brief one-sentence movie summary                                                         |   
| Director            | The name of the film's director                                                          |
| Actors              | A comma-separated list of the main stars of the film                                     |
| Year                | The year that the film released as an integer.                                           |
| Runtime (Minutes)   | The duration of the film in minutes                                                      |
| Rating              | User rating for the movie 0-10                                                           |
| Votes               | Number of votes                                                                          |
| Revenue (Millions)  | Movie revenue in millions                                                                |
| Metascore           | An aggregated average of critic scores (0-100), higher scores represent positive reviews |

#### Importing Packages

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 
sns.set()

from subprocess import check_output

#### Importing the dataset

In [39]:
movie_data = pd.read_csv(r"C:\Users\yjoshi\PythonCodes\DataSet\1000 movies data.csv")

In [11]:
movie_data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

## 3. Data Profiling

#### Understanding the DataSet

In [13]:
movie_data.shape

(1000, 12)

In [16]:
movie_data.isna().sum()

Rank                    0
Title                   0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

In [19]:
movie_data.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

In [20]:
movie_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
Rank                  1000 non-null int64
Title                 1000 non-null object
Genre                 1000 non-null object
Description           1000 non-null object
Director              1000 non-null object
Actors                1000 non-null object
Year                  1000 non-null int64
Runtime (Minutes)     1000 non-null int64
Rating                1000 non-null float64
Votes                 1000 non-null int64
Revenue (Millions)    872 non-null float64
Metascore             936 non-null float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.8+ KB


From above, we can see that __Metascore__ and __Revenue (Millions)__ column contain maximum null values. We will need to find proper value to replace nulls. As both these columns solely depends on movies performance and viewer's choice. 

### Pre-Profiling

In [21]:
import pandas_profiling
profile = pandas_profiling.ProfileReport(movie_data)
profile.to_file(outputfile="movie_data_preprocessing.html")

ModuleNotFoundError: No module named 'pandas_profiling'

### Pre-Processing

Handling missing data:
1. For missing Metascore values, since it is related to Rating we can replace missing Metascore values with Rating*10 values
2. For Revenue (Millions) missing values, we need to explore a little more as to which values to replace null with

In [88]:
movie_data.Metascore.fillna(movie_data.Rating*10,inplace=True)

In [23]:
movie_data.dtypes

Rank                    int64
Title                  object
Genre                  object
Description            object
Director               object
Actors                 object
Year                    int64
Runtime (Minutes)       int64
Rating                float64
Votes                   int64
Revenue (Millions)    float64
Metascore             float64
dtype: object

In [29]:
movie_data.Genre.unique().shape

(207,)

There are 207 unique Genres but this number is due to combinations of genres. Let us split this and get unique genre as the first mentioned Genre is the major genre the film belongs to.

#### Adding new columns to simplify data

In [40]:
movie_data[['Genre1','Genre2','Genre3']] = movie_data.Genre.str.split(",",expand=True,)

Lets fill the null values in Genre2 and Genre3 with Genre1 value

In [41]:
movie_data["Genre2"].fillna(movie_data.Genre1,inplace=True)

In [43]:
movie_data["Genre3"].fillna(movie_data.Genre2,inplace=True)

In [48]:
print(movie_data.Genre1.unique().shape)
print(movie_data.Genre2.unique().shape)
print(movie_data.Genre3.unique().shape)

(13,)
(19,)
(19,)


#### There are two columns representing rating, Metascore is on a scale of 0-100 and Rating is on a scale of 0-10, lets create a new column with average of both on a scale of 10.

In [91]:
movie_data["Avg_Rating"] = (movie_data["Rating"]+movie_data["Metascore"]/10)/2

In [90]:
movie_data

Rank                                    Title  \
0       1                  Guardians of the Galaxy   
1       2                               Prometheus   
2       3                                    Split   
3       4                                     Sing   
4       5                            Suicide Squad   
5       6                           The Great Wall   
6       7                               La La Land   
7       8                                 Mindhorn   
8       9                       The Lost City of Z   
9      10                               Passengers   
10     11  Fantastic Beasts and Where to Find Them   
11     12                           Hidden Figures   
12     13                                Rogue One   
13     14                                    Moana   
14     15                                 Colossal   
15     16                  The Secret Life of Pets   
16     17                            Hacksaw Ridge   
17     18                             Jason Bourne   
18     19                                     Lion   
19     20                                  Arrival   
20     21                                     Gold   
21     22                    Manchester by the Sea   
22     23                           Hounds of Love   
23     24                                   Trolls   
24     25             Independence Day: Resurgence   
25     26                          Paris pieds nus   
26     27                  Bahubali: The Beginning   
27     28                               Dead Awake   
28     29                                 Bad Moms   
29     30                         Assassin's Creed   
..    ...                                      ...   
970   971                        Texas Chainsaw 3D   
971   972                                Disturbia   
972   973                             Rock of Ages   
973   974                                 Scream 4   
974   975                           Queen of Katwe   
975   976               My Big Fat Greek Wedding 2   
976   977                              Dark Places   
977   978                            Amateur Night   
978   979           It's Only the End of the World   
979   980                       The Skin I Live In   
980   981                     Miracles from Heaven   
981   982                                    Annie   
982   983                      Across the Universe   
983   984                            Let's Be Cops   
984   985                                      Max   
985   986                            Your Highness   
986   987                      Final Destination 5   
987   988                             Endless Love   
988   989                                  Martyrs   
989   990                                    Selma   
990   991           Underworld: Rise of the Lycans   
991   992                         Taare Zameen Par   
992   993                     Take Me Home Tonight   
993   994                 Resident Evil: Afterlife   
994   995                                Project X   
995   996                     Secret in Their Eyes   
996   997                          Hostel: Part II   
997   998                   Step Up 2: The Streets   
998   999                             Search Party   
999  1000                               Nine Lives   

                          Genre               Director  \
0       Action,Adventure,Sci-Fi             James Gunn   
1      Adventure,Mystery,Sci-Fi           Ridley Scott   
2               Horror,Thriller     M. Night Shyamalan   
3       Animation,Comedy,Family   Christophe Lourdelet   
4      Action,Adventure,Fantasy             David Ayer   
5      Action,Adventure,Fantasy            Yimou Zhang   
6            Comedy,Drama,Music        Damien Chazelle   
7                        Comedy             Sean Foley   
8    Action,Adventure,Biography             James Gray   
9       Adventure,Drama,Romance          Morten Tyldum   
10     Adventure,F

In [59]:
print(movie_data.Votes.min())
print(movie_data.Votes.max())

movie_data[movie_data.Votes>1000000].Votes.count()

61
1791916


6

In [ ]:
movie_data = movie_data.drop(["Description"],axis=1)

In [100]:
movie_data.groupby("Director").filter(lambda x: len(x) >= 2).isnull().sum()

Rank                   0
Title                  0
Genre                  0
Director               0
Actors                 0
Year                   0
Runtime (Minutes)      0
Rating                 0
Votes                  0
Revenue (Millions)    26
Metascore              0
Genre1                 0
Genre2                 0
Genre3                 0
Avg_Rating             0
dtype: int64